lib/src/importer/node_package.dart

```diff
@@ -0,0 +1,388 @@
+// Copyright 2024 Google Inc. Use of this source code is governed by an
+// MIT-style license that can be found in the LICENSE file or at
+// https://opensource.org/licenses/MIT.
+
+import 'package:cli_pkg/js.dart';
+import 'package:collection/collection.dart';
+import 'package:sass/src/util/map.dart';
+import 'package:sass/src/util/nullable.dart';
+
+import '../importer.dart';
+import './utils.dart';
+import 'dart:convert';
+import '../io.dart';
+import 'package:path/path.dart' as p;
+
+/// An [Importer] that resolves `pkg:` URLs using the Node resolution algorithm.
+class NodePackageImporter extends Importer {
+  /// The starting path for canonicalizations without a containing URL.
+  late final String _entryPointDirectory;
+
+  /// Creates a Node package importer with the associated entry point.
+  NodePackageImporter(String? entryPointDirectory) {
+    if (entryPointDirectory == null) {
+      throw "The Node package importer cannot determine an entry point "
+          "because `require.main.filename` is not defined. "
+          "Please provide an `entryPointDirectory` to the `NodePackageImporter`.";
+    } else if (isBrowser) {
+      throw "The Node package importer cannot be used without a filesystem.";
+    }
+    _entryPointDirectory = p.absolute(entryPointDirectory);
+  }
+
+  @override
+  bool isNonCanonicalScheme(String scheme) => scheme == 'pkg';
+
+  @override
+  Uri? canonicalize(Uri url) {
+    if (url.scheme == 'file') return FilesystemImporter.cwd.canonicalize(url);
+    if (url.scheme != 'pkg') return null;
+
+    if (url.hasAuthority) {
+      throw "A pkg: URL must not have a host, port, username or password.";
+    } else if (p.url.isAbsolute(url.path)) {
+      throw "A pkg: URL's path must not begin with /.";
+    } else if (url.path.isEmpty) {
+      throw "A pkg: URL must not have an empty path.";
+    } else if (url.hasQuery || url.hasFragment) {
+      throw "A pkg: URL must not have a query or fragment.";
+    }
+
+    var basePath = containingUrl?.scheme == 'file'
+        ? p.fromUri(containingUrl!)
```

Comment:
This still needs a `p.dirname()`, since `containingUrl` points to a file and not a directory.

Reply:
That change was made in https://github.com/sass/dart-sass/pull/2130/commits/53472fdb880494924f06d5b0974cfac94b36b433